TransferVersionModule fails when tickets are linked across projects

## The Bug
The TransferVersion module fails when a ticket is marked as duplicate of a ticket from another project, because the version it tries to add does not exist in the parent project.

## Example Ticket
[BDS-4621](https://bugs.mojang.com/browse/BDS-4621)

## Log message
```
[RESPONSE] [BDS-4621] [TransferVersions] Failed
net.rcarz.jiraclient.JiraException: Failed to update issue MCPE-67169
    at net.rcarz.jiraclient.Issue$FluentUpdate.execute(Issue.java:380)
    at io.github.mojira.arisa.infrastructure.JiraKt$addAffectedVersionById$1$1.invokeSuspend(Jira.kt:139)
    at io.github.mojira.arisa.infrastructure.JiraKt$addAffectedVersionById$1$1.invoke(Jira.kt)
    at arrow.core.Either$Companion.catch(Either.kt:862)
    at arrow.core.Either$Companion.catch(Either.kt:858)
    at io.github.mojira.arisa.infrastructure.JiraKt$addAffectedVersionById$1.invokeSuspend(Jira.kt:135)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:271)
    at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
    at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
    at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
    at io.github.mojira.arisa.infrastructure.JiraKt.addAffectedVersionById(Jira.kt:134)
    at io.github.mojira.arisa.ModuleRegistry$9$1$1$2$2.invoke(ModuleRegistry.kt:285)
    at io.github.mojira.arisa.ModuleRegistry$9$1$1$2$2.invoke(ModuleRegistry.kt:62)
    at arrow.syntax.function.PartialsKt$partially1$2.invoke(partials.kt:773)
    at arrow.syntax.function.PartialsKt$partially1$1.invoke(partials.kt:771)
    at io.github.mojira.arisa.modules.ModuleKt.tryRunAll(Module.kt:48)
    at io.github.mojira.arisa.modules.TransferVersionsModule$invoke$$inlined$with$lambda$1.invokeSuspend(TransferVersionsModule.kt:36)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at arrow.typeclasses.MonadContinuation$bind$$inlined$suspendCoroutineUninterceptedOrReturn$lambda$1.invoke(MonadContinuations.kt:36)
    at arrow.typeclasses.MonadContinuation$bind$$inlined$suspendCoroutineUninterceptedOrReturn$lambda$1.invoke(MonadContinuations.kt:15)
    at arrow.core.extensions.EitherMonad$flatMap$1.invoke(either.kt:123)
    at arrow.core.extensions.EitherMonad$flatMap$1.invoke(either.kt:115)
    at arrow.core.EitherKt.flatMap(Either.kt:881)
    at arrow.core.extensions.EitherMonad$DefaultImpls.flatMap(either.kt:123)
    at arrow.core.extensions.either.monad.EitherMonadKt$monad_singleton$1.flatMap(EitherMonad.kt:21)
    at arrow.core.extensions.either.monad.EitherMonadKt$monad_singleton$1.flatMap(EitherMonad.kt:21)
    at arrow.typeclasses.MonadContinuation.flatMap(MonadContinuations.kt)
    at arrow.typeclasses.MonadContinuation.bind$suspendImpl(MonadContinuations.kt:34)
    at arrow.typeclasses.MonadContinuation.bind(MonadContinuations.kt)
    at io.github.mojira.arisa.modules.TransferVersionsModule$invoke$$inlined$with$lambda$1.invokeSuspend(TransferVersionsModule.kt:34)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at arrow.typeclasses.MonadContinuation$bind$$inlined$suspendCoroutineUninterceptedOrReturn$lambda$1.invoke(MonadContinuations.kt:36)
    at arrow.typeclasses.MonadContinuation$bind$$inlined$suspendCoroutineUninterceptedOrReturn$lambda$1.invoke(MonadContinuations.kt:15)
    at arrow.core.extensions.EitherMonad$flatMap$1.invoke(either.kt:123)
    at arrow.core.extensions.EitherMonad$flatMap$1.invoke(either.kt:115)
    at arrow.core.EitherKt.flatMap(Either.kt:881)
    at arrow.core.extensions.EitherMonad$DefaultImpls.flatMap(either.kt:123)
    at arrow.core.extensions.either.monad.EitherMonadKt$monad_singleton$1.flatMap(EitherMonad.kt:21)
    at arrow.core.extensions.either.monad.EitherMonadKt$monad_singleton$1.flatMap(EitherMonad.kt:21)
    at arrow.typeclasses.MonadContinuation.flatMap(MonadContinuations.kt)
    at arrow.typeclasses.MonadContinuation.bind$suspendImpl(MonadContinuations.kt:34)
    at arrow.typeclasses.MonadContinuation.bind(MonadContinuations.kt)
    at io.github.mojira.arisa.modules.TransferVersionsModule$invoke$$inlined$with$lambda$1.invokeSuspend(TransferVersionsModule.kt:29)
    at io.github.mojira.arisa.modules.TransferVersionsModule$invoke$$inlined$with$lambda$1.invoke(TransferVersionsModule.kt)
    at arrow.typeclasses.MonadFx$monad$wrapReturn$1.invokeSuspend(Monad.kt:108)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:128)
    at arrow.typeclasses.MonadFx$DefaultImpls.monad(Monad.kt:109)
    at arrow.core.extensions.EitherMonadFx.monad(either.kt:136)
    at arrow.core.extensions.EitherMonadFx.monad(either.kt:133)
    at arrow.core.extensions.EitherKt.fx(either.kt:235)
    at io.github.mojira.arisa.modules.TransferVersionsModule.invoke(TransferVersionsModule.kt:27)
    at io.github.mojira.arisa.modules.TransferVersionsModule.invoke(TransferVersionsModule.kt:9)
    at io.github.mojira.arisa.ModuleRegistry$register$2$1.invoke(ModuleRegistry.kt:86)
    at io.github.mojira.arisa.ModuleRegistry$register$2$1.invoke(ModuleRegistry.kt:62)
    at io.github.mojira.arisa.ModuleRegistry.tryExecuteModule(ModuleRegistry.kt:90)
    at io.github.mojira.arisa.ModuleRegistry.access$tryExecuteModule(ModuleRegistry.kt:62)
    at io.github.mojira.arisa.ModuleRegistry$register$2.invoke(ModuleRegistry.kt:86)
    at io.github.mojira.arisa.ModuleRegistry$register$2.invoke(ModuleRegistry.kt:62)
    at arrow.syntax.function.PartialsKt$partially2$1.invoke(partials.kt:775)
    at io.github.mojira.arisa.ModuleExecutor.executeModule(ModuleExecutor.kt:81)
    at io.github.mojira.arisa.ModuleExecutor.execute(ModuleExecutor.kt:33)
    at io.github.mojira.arisa.ArisaMainKt.main(ArisaMain.kt:51)
    at io.github.mojira.arisa.ArisaMainKt.main(ArisaMain.kt)
Caused by: net.rcarz.jiraclient.RestException: 400 : {"errorMessages":[],"errors":{"versions":"Version id '19406' is not valid"}}
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:171)
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:193)
    at net.rcarz.jiraclient.RestClient.request(RestClient.java:238)
    at net.rcarz.jiraclient.RestClient.put(RestClient.java:434)
    at net.rcarz.jiraclient.RestClient.put(RestClient.java:452)
    at net.rcarz.jiraclient.Issue$FluentUpdate.execute(Issue.java:378) 
```

--- a/src/main/kotlin/io/github/mojira/arisa/ModuleRegistry.kt
+++ b/src/main/kotlin/io/github/mojira/arisa/ModuleRegistry.kt
@@ -275,7 +275,10 @@ class ModuleRegistry(jiraClient: JiraClient, private val config: Config) {
                         TransferVersionsModule.Link(
                             link.type.name,
                             link.outwardIssue != null,
-                            link.outwardIssue.key
+                            if (link.outwardIssue != null)
+                                link.outwardIssue.key
+                            else
+                                link.inwardIssue.key
                         ) {
                             if (link.outwardIssue != null) {
                                 getIssue(jiraClient, link.outwardIssue.key)

--- a/src/main/kotlin/io/github/mojira/arisa/ModuleRegistry.kt
+++ b/src/main/kotlin/io/github/mojira/arisa/ModuleRegistry.kt
@@ -269,11 +269,13 @@ class ModuleRegistry(jiraClient: JiraClient, private val config: Config) {
             TransferVersionsModule()
         ) { issue ->
             TransferVersionsModule.Request(
+                issue.key,
                 issue.issueLinks
                     .map { link ->
                         TransferVersionsModule.Link(
                             link.type.name,
-                            link.outwardIssue != null
+                            link.outwardIssue != null,
+                            link.outwardIssue.key
                         ) {
                             if (link.outwardIssue != null) {
                                 getIssue(jiraClient, link.outwardIssue.key)

--- a/src/main/kotlin/io/github/mojira/arisa/modules/TransferVersionsModule.kt
+++ b/src/main/kotlin/io/github/mojira/arisa/modules/TransferVersionsModule.kt
@@ -15,17 +15,22 @@ class TransferVersionsModule : Module<TransferVersionsModule.Request> {
     data class Link(
         val type: String,
         val outwards: Boolean,
+        val issueKey: String,
         val getLinkedIssue: () -> Either<Throwable, LinkedIssue>
     )
 
     data class Request(
+        val key: String,
         val links: List<Link>,
         val versions: List<String>
     )
 
     override fun invoke(request: Request): Either<ModuleError, ModuleResponse> = with(request) {
         Either.fx {
-            val parentLinks = links.filter(::isDuplicatesLink)
+            val parentLinks = links
+                .filter(::isDuplicatesLink)
+                .filter(::isSameProject.partially2(key))
+
             assertGreaterThan(parentLinks.size, 0).bind()
             val parents = parentLinks
                 .map { it.getLinkedIssue() }
@@ -39,6 +44,12 @@ class TransferVersionsModule : Module<TransferVersionsModule.Request> {
 
     private fun isDuplicatesLink(link: Link) =
         link.type.toLowerCase() == "duplicate" && link.outwards
+
+    private fun isSameProject(link: Link, key: String) =
+        link.issueKey.getProject() == key.getProject()
+
+    private fun String.getProject() =
+        substring(0, indexOf('-'))
 
     private fun applyParents(version: String, parents: List<Either<Throwable, LinkedIssue>>) =
         parents.map { parentEither ->

--- a/src/test/kotlin/io/github/mojira/arisa/modules/TransferVersionsTest.kt
+++ b/src/test/kotlin/io/github/mojira/arisa/modules/TransferVersionsTest.kt
@@ -15,7 +15,7 @@ import io.kotest.matchers.shouldBe
 class TransferVersionsTest : StringSpec({
     "should return OperationNotNeededModuleResponse when there are no issue links" {
         val module = TransferVersionsModule()
-        val request = Request(emptyList(), listOf("v1"))
+        val request = Request("MC-1", emptyList(), listOf("v1"))
 
         val result = module(request)
 
@@ -24,8 +24,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return OperationNotNeededModuleResponse when there is no duplicates link" {
         val module = TransferVersionsModule()
-        val link = Link("Relates", true) { LinkedIssue(emptyList()) { Unit.right() }.right() }
-        val request = Request(listOf(link), listOf("v1"))
+        val link = Link("Relates", true, "MC-1") { LinkedIssue(emptyList()) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
 
         val result = module(request)
 
@@ -34,8 +34,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return OperationNotNeededModuleResponse when there is no outgoing duplicates link" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", false) { LinkedIssue(emptyList()) { Unit.right() }.right() }
-        val request = Request(listOf(link), listOf("v1"))
+        val link = Link("Duplicate", false, "MC-1") { LinkedIssue(emptyList()) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
 
         val result = module(request)
 
@@ -44,8 +44,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return OperationNotNeededModuleResponse when the issue has no affected versions" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { LinkedIssue(emptyList()) { Unit.right() }.right() }
-        val request = Request(listOf(link), emptyList())
+        val link = Link("Duplicate", true, "MC-1") { LinkedIssue(emptyList()) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), emptyList())
 
         val result = module(request)
 
@@ -54,8 +54,18 @@ class TransferVersionsTest : StringSpec({
 
     "should return OperationNotNeededModuleResponse when the parent already has all versions" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { LinkedIssue(listOf("v1")) { Unit.right() }.right() }
-        val request = Request(listOf(link), listOf("v1"))
+        val link = Link("Duplicate", true, "MC-1") { LinkedIssue(listOf("v1")) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
+
+        val result = module(request)
+
+        result.shouldBeLeft(OperationNotNeededModuleResponse)
+    }
+
+    "should return OperationNotNeededModuleResponse when the parent is from a different project" {
+        val module = TransferVersionsModule()
+        val link = Link("Duplicate", true, "MCL-1") { LinkedIssue(emptyList()) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
 
         val result = module(request)
 
@@ -64,8 +74,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return ModuleResponse when there are versions to transfer" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { LinkedIssue(emptyList()) { Unit.right() }.right() }
-        val request = Request(listOf(link), listOf("v1"))
+        val link = Link("Duplicate", true, "MC-1") { LinkedIssue(emptyList()) { Unit.right() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
 
         val result = module(request)
 
@@ -76,7 +86,7 @@ class TransferVersionsTest : StringSpec({
         var firstVersionAdded = false
         var secondVersionAdded = false
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) {
+        val link = Link("Duplicate", true, "MC-1") {
             LinkedIssue(emptyList()) { v ->
                 when (v) {
                     "v1" -> firstVersionAdded = true
@@ -85,7 +95,7 @@ class TransferVersionsTest : StringSpec({
                 Unit.right()
             }.right()
         }
-        val request = Request(listOf(link), listOf("v1", "v2"))
+        val request = Request("MC-1", listOf(link), listOf("v1", "v2"))
 
         val result = module(request)
 
@@ -98,19 +108,19 @@ class TransferVersionsTest : StringSpec({
         var addedToFirstParent = false
         var addedToSecondParent = false
         val module = TransferVersionsModule()
-        val link1 = Link("Duplicate", true) {
+        val link1 = Link("Duplicate", true, "MC-1") {
             addedToFirstParent = true
             LinkedIssue(emptyList()) {
                 Unit.right()
             }.right()
         }
-        val link2 = Link("Duplicate", true) {
+        val link2 = Link("Duplicate", true, "MC-1") {
             addedToSecondParent = true
             LinkedIssue(emptyList()) {
                 Unit.right()
             }.right()
         }
-        val request = Request(listOf(link1, link2), listOf("v1"))
+        val request = Request("MC-1", listOf(link1, link2), listOf("v1"))
 
         val result = module(request)
 
@@ -121,8 +131,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return FailedModuleResponse when adding a version fails" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { LinkedIssue(emptyList()) { RuntimeException().left() }.right() }
-        val request = Request(listOf(link), listOf(""))
+        val link = Link("Duplicate", true, "MC-1") { LinkedIssue(emptyList()) { RuntimeException().left() }.right() }
+        val request = Request("MC-1", listOf(link), listOf(""))
 
         val result = module(request)
 
@@ -133,8 +143,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return FailedModuleResponse with all errors when adding multiple versions fails" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { LinkedIssue(emptyList()) { RuntimeException().left() }.right() }
-        val request = Request(listOf(link), listOf("v1", "v2"))
+        val link = Link("Duplicate", true, "MC-1") { LinkedIssue(emptyList()) { RuntimeException().left() }.right() }
+        val request = Request("MC-1", listOf(link), listOf("v1", "v2"))
 
         val result = module(request)
 
@@ -145,8 +155,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return FailedModuleResponse when getting an issue fails" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { RuntimeException().left() }
-        val request = Request(listOf(link), listOf("v1"))
+        val link = Link("Duplicate", true, "MC-1") { RuntimeException().left() }
+        val request = Request("MC-1", listOf(link), listOf("v1"))
 
         val result = module(request)
 
@@ -157,8 +167,8 @@ class TransferVersionsTest : StringSpec({
 
     "should return FailedModuleResponse with all errors when getting an issue fails" {
         val module = TransferVersionsModule()
-        val link = Link("Duplicate", true) { RuntimeException().left() }
-        val request = Request(listOf(link, link), listOf("v1"))
+        val link = Link("Duplicate", true, "MC-1") { RuntimeException().left() }
+        val request = Request("MC-1", listOf(link, link), listOf("v1"))
 
         val result = module(request)
 

